kms/azurekms/utils.go

```diff
@@ -101,23 +104,78 @@ func parseKeyName(rawURI string, defaults defaultOptions) (vault, name, version
 }
 
 func convertKey(key *azkeys.JSONWebKey) (crypto.PublicKey, error) {
-	// Hack to be able to properly convert keys in HSM.
-	if key != nil && key.Kty != nil {
-		switch *key.Kty {
-		case azkeys.JSONWebKeyTypeECHSM:
-			key.Kty = pointer(azkeys.JSONWebKeyTypeEC)
-		case azkeys.JSONWebKeyTypeRSAHSM:
-			key.Kty = pointer(azkeys.JSONWebKeyTypeRSA)
-		}
+	if key == nil || key.Kty == nil {
+		return nil, errors.New("invalid key: missing kty value")
 	}
 
-	b, err := json.Marshal(key)
-	if err != nil {
-		return nil, errors.Wrap(err, "error marshaling key")
+	switch *key.Kty {
+	case azkeys.JSONWebKeyTypeEC, azkeys.JSONWebKeyTypeECHSM:
+		return ecPublicKey(key.Crv, key.X, key.Y)
+	case azkeys.JSONWebKeyTypeRSA, azkeys.JSONWebKeyTypeRSAHSM:
+		return rsaPublicKey(key.N, key.E)
+	case azkeys.JSONWebKeyTypeOct, azkeys.JSONWebKeyTypeOctHSM:
+		return octPublicKey(key.K)
+	default:
+		return nil, fmt.Errorf("invalid key: unsupported kty '%s'", *key.Kty)
+	}
+}
+
+func ecPublicKey(crv *azkeys.JSONWebKeyCurveName, x, y []byte) (crypto.PublicKey, error) {
+	if crv == nil {
+		return nil, errors.New("invalid EC key: missing crv value")
+	}
+	if len(x) == 0 || len(y) == 0 {
+		return nil, errors.New("invalid EC key: missing x or y values")
+	}
+
+	var curve elliptic.Curve
+	var curveSize int
+	switch *crv {
+	case azkeys.JSONWebKeyCurveNameP256:
+		curve = elliptic.P256()
+		curveSize = 32
+	case azkeys.JSONWebKeyCurveNameP384:
+		curve = elliptic.P384()
+		curveSize = 48
+	case azkeys.JSONWebKeyCurveNameP521:
+		curve = elliptic.P521()
+		curveSize = 66 // (521/8 + 1)
+	case azkeys.JSONWebKeyCurveNameP256K:
+		return nil, errors.New("invalid EC key: crv 'P-256K' is not supported")
+	default:
+		return nil, fmt.Errorf("invalid EC key: crv '%s' is not supported", *crv)
+	}
+
+	if len(x) != curveSize || len(y) != curveSize {
+		return nil, errors.New("invalid EC key: x or y length is not valid")
 	}
-	var jwk jose.JSONWebKey
-	if err := jwk.UnmarshalJSON(b); err != nil {
-		return nil, errors.Wrap(err, "error unmarshaling key")
+
+	key := &ecdsa.PublicKey{
+		Curve: curve,
+		X:     new(big.Int).SetBytes(x),
+		Y:     new(big.Int).SetBytes(y),
+	}
+
+	if !curve.IsOnCurve(key.X, key.Y) {
+		return nil, errors.New("invalid EC key: point (x, y) does not lie on the curve")
+	}
+
+	return key, nil
+}
+
+func rsaPublicKey(n, e []byte) (crypto.PublicKey, error) {
+	if len(n) == 0 || len(e) == 0 {
+		return nil, errors.New("invalid RSA key: missing n or e values")
+	}
+	return &rsa.PublicKey{
+		N: new(big.Int).SetBytes(n),
+		E: int(new(big.Int).SetBytes(e).Int64()),
+	}, nil
+}
+
+func octPublicKey(k []byte) (crypto.PublicKey, error) {
+	if k == nil {
+		return nil, errors.New("invalid oct key: missing k value")
 	}
```

Comment:
For the others you test the `len()` of the inputs. Maybe good to make it consistent.

Reply:
And empty password is valid password, but not for other types.